src/Connection/DefaultConnectionPool.php

```diff
@@ -82,141 +95,171 @@ public function getStream(Request $request, CancellationToken $cancellation): Pr
-
-                        // In case of a user cancellation request, throw the expected exception
-                        $cancellation->throwIfRequested();
-
-                        // Otherwise we ran into a timeout of our TimeoutCancellationToken
-                        throw new TimeoutException(\sprintf(
-                            "TLS handshake with '%s' @ '%s' timed out, took longer than " . $request->getTlsHandshakeTimeout() . ' ms',
-                            $authority,
-                            $socket->getRemoteAddress()->toString()
-                        )); // don't pass $e
-                    }
-
-                    $tlsInfo = $socket->getTlsInfo();
-                    \assert($tlsInfo !== null);
-
-                    if ($tlsInfo->getApplicationLayerProtocol() === 'h2') {
-                        $connection = new Http2Connection($socket);
-                        yield $connection->initialize();
-                    } else {
-                        if (!\array_intersect($request->getProtocolVersions(), ['1.0', '1.1'])) {
-                            $socket->close();
-                            throw new HttpException('Downgrade to HTTP/1.x forbidden, but server does not support HTTP/2');
-                        }
-
-                        $connection = new Http1Connection($socket);
-                    }
-                } else {
-                    $connection = new Http1Connection($socket);
-                }
-
-                \assert($promise instanceof Promise);
-
-                $connection->onClose(function () use ($key, $promise) {
-                    $this->connections[$key]->detach($promise);
-
-                    if (!$this->connections[$key]->count()) {
-                        unset($this->connections[$key]);
-                    }
-                });
+                return $connection->getStream($request);
+            }
 
-                return $connection;
-            });
+            $promise = new Coroutine($this->createConnection($request, $cancellation, $authority, $isHttps));
 
-            $this->connections[$key] = $this->connections[$key] ?? new \SplObjectStorage;
-            $this->connections[$key]->attach($promise);
+            $hash = \spl_object_hash($promise);
+            $this->connections[$key] = $this->connections[$key] ?? [];
+            $this->connections[$key][$hash] = $promise;
 
             try {
                 $connection = yield $promise;
                 \assert($connection instanceof Connection);
             } catch (\Throwable $exception) {
                 // Connection failed, remove from list of connections.
-                $this->connections[$key]->detach($promise);
+                unset($this->connections[$key][$hash]);
 
-                if (!$this->connections[$key]->count()) {
+                if (empty($this->connections[$key])) {
                     unset($this->connections[$key]);
                 }
```

Comment:
This could be extracted into a `dropConnection` method, as it's duplicated below.